File: src/util/pace.ts
/**
 * Usage pace calculation.
 *
 * Computes how a user's actual usage compares to a linear "even burn"
 * expectation across a rate-limit window. Ported from CodexBar's
 * `UsagePace.swift` and Win-CodexBar's `usage_pace.rs`.
 *
 * The math:
 *   elapsed  = windowDuration - timeUntilReset
 *   expected = (elapsed / windowDuration) * 100
 *   delta    = actualUsedPercent - expected
 *
 * A positive delta means the user is burning faster than the linear
 * rate (ahead of pace / at risk of hitting the cap early). A negative
 * delta means they're burning slower (behind pace / will have headroom
 * left when the window resets).
 *
 * Threshold bands (matching CodexBar):
 *   |delta| <= 2  → "On pace"
 *   delta  < -2   → "Behind (-X%)"   — under budget, good
 *   delta  >  2   → "Ahead (+X%)"    — over budget, risky
 */

export interface PaceResult {
  /** Signed delta: positive = ahead (burning fast), negative = behind (safe). */
  delta: number;
  /** Human label: "On pace", "Behind (-12%)", "Ahead (+8%)". */
  label: string;
  /** Qualitative band for optional color coding. */
  band: "on-pace" | "behind" | "ahead";
}

const ON_PACE_THRESHOLD = 2;

/**
 * Compute pace for a rate-limit window.
 *
 * @param usedPercent   How much of the window has been consumed (0–100).
 * @param resetInSeconds  Seconds until the window resets.
 * @param windowSeconds   Total window duration in seconds (e.g. 18000 for 5h, 604800 for 7d).
 * @returns PaceResult, or undefined if inputs are missing / nonsensical.
 */
export function computePace(
  usedPercent: number | undefined,
  resetInSeconds: number | undefined,
  windowSeconds: number,
): PaceResult | undefined {
  if (
    typeof usedPercent !== "number" ||
    typeof resetInSeconds !== "number" ||
    windowSeconds <= 0
  ) {
    return undefined;
  }

  const elapsed = Math.max(0, windowSeconds - resetInSeconds);
  const expected = (elapsed / windowSeconds) * 100;
  const actual = Math.max(0, Math.min(100, usedPercent));
  const delta = actual - expected;
  const absDelta = Math.abs(delta);
  const rounded = Math.round(absDelta);

  if (absDelta <= ON_PACE_THRESHOLD) {
    return { delta: 0, label: "On pace", band: "on-pace" };
  }

  if (delta < 0) {
    return {
      delta: Math.round(delta),
      label: `Behind (-${rounded}%)`,
      band: "behind",
    };
  }

  return {
    delta: Math.round(delta),
    label: `Ahead (+${rounded}%)`,
    band: "ahead",
  };
}

File: internal/providers/pace.go
package providers

import (
	"fmt"
	"math"
	"time"
)

// PaceInput is the data a provider passes to compute a pace metric.
type PaceInput struct {
	MetricID       string        // e.g. "session-pace", "weekly-pace"
	Label          string        // e.g. "Session", "Weekly"
	Name           string        // e.g. "Session pace"
	UsedPercent    float64       // 0-100, how much has been consumed
	WindowDuration time.Duration // total window length
	ResetIn        time.Duration // time remaining until reset
}

// PaceMetric computes a linear-burn-rate pace metric.
//
// Returns nil if the input is degenerate (zero window, no elapsed time).
//
// The metric value shows the delta between actual and expected usage:
//   - Positive value = reserve (used less than expected — good)
//   - Negative value = deficit (used more than expected — bad)
//
// Pace metrics intentionally do NOT set a Ratio: mapping ±N% delta
// onto a 0..1 fill bar produces a "half-filled tile" at 0% that reads
// like a percent meter at 50% and confuses everyone. The signed value
// + caption ("On pace" / "Reserve" / "Deficit") carries the signal
// without a misleading fill.
func PaceMetric(in PaceInput) *MetricValue {
	windowSec := in.WindowDuration.Seconds()
	resetSec := in.ResetIn.Seconds()
	if windowSec <= 0 {
		return nil
	}
	if resetSec < 0 {
		resetSec = 0
	}

	elapsed := windowSec - resetSec
	if elapsed <= 0 {
		return nil // window just started, no data point
	}

	expectedUsed := (elapsed / windowSec) * 100
	actualUsed := math.Max(0, math.Min(100, in.UsedPercent))

	// delta > 0 means used MORE than expected (deficit)
	// delta < 0 means used LESS than expected (reserve)
	delta := actualUsed - expectedUsed

	// Flip sign so positive = reserve (good), negative = deficit (bad)
	reserve := -delta

	// Display value — always signed so it doesn't read as "percent
	// remaining." Near pace (< 1%) we drop to one-decimal precision
	// so e.g. "+0.3%" / "-0.7%" instead of a bare "0%".
	absReserve := math.Abs(reserve)
	var valueStr string
	switch {
	case absReserve < 0.05:
		valueStr = "±0.0%"
	case absReserve < 1:
		valueStr = fmt.Sprintf("%+.1f%%", reserve)
	default:
		valueStr = fmt.Sprintf("%+d%%", int(math.Round(reserve)))
	}

	// Caption disambiguates a pace tile from its percent sibling
	// (same top label, e.g. "SESSION"). The +/- on the value already
	// conveys direction, so a single constant word is enough.
	caption := "Pace"

	nv := reserve
	// Intentionally no ResetInSeconds: the renderer's subvalue priority
	// puts countdown above caption, which would hide the pace signal
	// ("On pace" / "Reserve" / "Deficit") behind a timer that already
	// shows on the sibling percent tile.
	return &MetricValue{
		ID:              in.MetricID,
		Label:           in.Label,
		Name:            in.Name,
		Value:           valueStr,
		NumericValue:    &nv,
		NumericUnit:     "percent",
		NumericGoodWhen: "high",
		Caption:         caption,
		UpdatedAt:       time.Now().UTC().Format(time.RFC3339),
	}
}
